File: strategy_core/queries.py
"""Typed query objects for the grouped strategy data client."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from strategy_core.minutetemp import OracleRankBy, OracleScoreMode, ReportType

DateLike = datetime | str
LocalDateLike = date | str


@dataclass(frozen=True, slots=True)
class LimitsQuery:
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class ForecastQuery:
    model_id: str | None = None
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class OracleScoresQuery:
    days: str = "7"
    mode: OracleScoreMode | str = "day_ahead"
    rank_by: OracleRankBy | str = "high"
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class ForecastRunsQuery:
    model_id: str | None = None
    start: DateLike | None = None
    end: DateLike | None = None
    limit: int | None = None
    cursor: str | None = None
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class ForecastRunQuery:
    run_id: str
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class LatestReportsQuery:
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class ReportsQuery:
    report_type: ReportType | None = None
    date: LocalDateLike | None = None
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class ReportHistoryQuery:
    report_type: ReportType | None = None
    start: LocalDateLike | None = None
    end: LocalDateLike | None = None
    limit: int | None = None
    cursor: str | None = None
    refresh: bool = False


@dataclass(frozen=True, slots=True)
class LatestObservationQuery:
    refresh: bool = False

File: native/strategy_core/src/queries.rs
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum DateLike {
    DateTime(DateTime<Utc>),
    String(String),
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum LocalDateLike {
    Date(NaiveDate),
    String(String),
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct LimitsQuery {
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ForecastQuery {
    pub model_id: Option<String>,
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct OracleScoresQuery {
    #[serde(default = "default_oracle_days")]
    pub days: String,
    #[serde(default = "default_oracle_mode")]
    pub mode: String,
    #[serde(default = "default_oracle_rank_by")]
    pub rank_by: String,
    #[serde(default)]
    pub refresh: bool,
}

impl Default for OracleScoresQuery {
    fn default() -> Self {
        Self {
            days: default_oracle_days(),
            mode: default_oracle_mode(),
            rank_by: default_oracle_rank_by(),
            refresh: false,
        }
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ForecastRunsQuery {
    pub model_id: Option<String>,
    pub start: Option<DateLike>,
    pub end: Option<DateLike>,
    pub limit: Option<i64>,
    pub cursor: Option<String>,
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ForecastRunQuery {
    pub run_id: String,
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct LatestReportsQuery {
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReportsQuery {
    pub report_type: Option<String>,
    pub date: Option<LocalDateLike>,
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReportHistoryQuery {
    pub report_type: Option<String>,
    pub start: Option<LocalDateLike>,
    pub end: Option<LocalDateLike>,
    pub limit: Option<i64>,
    pub cursor: Option<String>,
    #[serde(default)]
    pub refresh: bool,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct LatestObservationQuery {
    pub day_mode: Option<String>,
    #[serde(default)]
    pub refresh: bool,
}

fn default_oracle_days() -> String {
    "7".to_string()
}

fn default_oracle_mode() -> String {
    "day_ahead".to_string()
}

fn default_oracle_rank_by() -> String {
    "high".to_string()
}
